torch/_functorch/partitioners.py

```diff
@@ -120,6 +125,46 @@ def _extract_fwd_bwd_modules(joint_module: fx.GraphModule, saved_values, saved_s
                     saved_sym_nodes.remove(saved_sym)
                     break
 
+    # Now that we have the finalized list of saved values, we need to ensure
+    # we propagate all symbols which are referenced by backwards inputs.
+    # These are not directly used in the graph but are required for downstream
+    # sizevar assignment
+    saved_symbols: Set[sympy.Symbol] = set()
+    saved_sym_nodes_binding = []
+    saved_sym_nodes_derived = []
+
+    # Some symbols may already be bound in the directly saved_sym_nodes,
+    # keep track of them so we don't re-bind them
+    for node in saved_sym_nodes:
+        symbol = is_symbol_binding_fx_node(node)
+        if symbol:
+            saved_symbols.add(symbol)
+            saved_sym_nodes_binding.append(node)
+        else:
+            saved_sym_nodes_derived.append(node)
+
+    # Now go through all of the prospective backward inputs and track any
+    # other symbols we need to bind
+    symbol_bindings = find_symbol_binding_fx_nodes(joint_module.graph)
+    for node in itertools.chain(saved_sym_nodes_derived, saved_values, tangent_inputs):
+        if "val" not in node.meta:
+            continue
+        new_symbols = free_symbols(node.meta["val"]) - saved_symbols
+        # NB: Deterministic order please!
+        for s in sorted(new_symbols, key=lambda s: s.name):
+            # NB: For well formed graphs, the symbol should always be present,
+            # but we also have ways to produce ill-formed graphs, e.g., direct
+            # make_fx usages, so don't choke in this case
+            if s not in symbol_bindings:
+                continue
+            saved_sym_nodes_binding.append(symbol_bindings[s])
+        saved_symbols |= new_symbols
```

Comment:
I agree that this is strictly better, but mostly just confirming: Instead of doing all of this pruning, could you just take every SymInt GraphArg from the forward, and force it to be saved for the backward?

I can buy the moral argument that we shouldn't bother saving things for backwards unnecessarily if we don't need them, although it probably wouldn't impact perf (since we're just saving sympy expressions / ints)

Reply:
Yes, it is permissible not to do pruning.  Less things saved for backwards actually does marginally improve perf, because stashing things on autograd Function has to be done at runtime. Individually it doesn't matter, but I've seen some graphs with hundreds of symbols... seems suspicious.